File: internal/handlers/cookie_analytics.go
package handlers

import (
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/bright-interaction/slab/internal/analyticsdb"
	"github.com/bright-interaction/slab/internal/config"
	"github.com/bright-interaction/slab/internal/store"
)

// CookieAnalyticsHandler exposes the stitched analytics surface that
// reads consent_records + visit_events + visit_sessions + visit_engagement
// through the DuckDB ATTACH on the SQLite file. All endpoints are
// authenticated admin routes scoped to a site.
//
// When analyticsMgr is nil (DuckDB failed to open at boot, or the build
// shipped without CGO), every endpoint returns 503 with a human-readable
// error so the dashboard can degrade gracefully instead of throwing.
type CookieAnalyticsHandler struct {
	cfg          *config.Config
	queries      *store.Queries
	analyticsMgr *analyticsdb.Manager
}

// NewCookieAnalyticsHandler constructs the handler. Pass nil for
// analyticsMgr when DuckDB is unavailable; all routes will return 503
// with a "stitched analytics unavailable" body in that case.
func NewCookieAnalyticsHandler(cfg *config.Config, queries *store.Queries, mgr *analyticsdb.Manager) *CookieAnalyticsHandler {
	return &CookieAnalyticsHandler{cfg: cfg, queries: queries, analyticsMgr: mgr}
}

// available is the precondition guard every endpoint runs first. Returns
// true when both the route's site_id is well-formed AND the manager is
// non-nil. On failure it has already written the response.
func (h *CookieAnalyticsHandler) available(w http.ResponseWriter, r *http.Request) (string, bool) {
	siteID := urlParam(r, "siteID")
	if !isSafeSiteID(siteID) {
		writeError(w, http.StatusBadRequest, "Invalid siteID")
		return "", false
	}
	if h.analyticsMgr == nil {
		writeError(w, http.StatusServiceUnavailable, "Stitched analytics unavailable: DuckDB layer not initialised. Check server logs.")
		return "", false
	}
	return siteID, true
}

// parseTimeWindow reads ?from=&to= as either RFC3339 or unix-ms. Default:
// last 30 days. Returns the same TimeRange the analyticsdb queries expect.
func parseTimeWindow(r *http.Request) analyticsdb.TimeRange {
	now := time.Now().UTC()
	from := now.AddDate(0, 0, -30)
	to := now
	if v := r.URL.Query().Get("from"); v != "" {
		if t, err := time.Parse(time.RFC3339, v); err == nil {
			from = t
		} else if n, err := strconv.ParseInt(v, 10, 64); err == nil {
			from = time.UnixMilli(n).UTC()
		}
	}
	if v := r.URL.Query().Get("to"); v != "" {
		if t, err := time.Parse(time.RFC3339, v); err == nil {
			to = t
		} else if n, err := strconv.ParseInt(v, 10, 64); err == nil {
			to = time.UnixMilli(n).UTC()
		}
	}
	return analyticsdb.NewTimeRange(from, to)
}

// Funnel returns the headline KPIs (visitors / consenting / accepted /
// rejected / GPC / abandoned) over the requested window.
//
// GET /api/sites/{siteID}/cookies/analytics/funnel?from=&to=
func (h *CookieAnalyticsHandler) Funnel(w http.ResponseWriter, r *http.Request) {
	siteID, ok := h.available(w, r)
	if !ok {
		return
	}
	tr := parseTimeWindow(r)
	kpis, err := h.analyticsMgr.FunnelKPIs(r.Context(), siteID, tr)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Failed to compute funnel: "+err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"site_id":      siteID,
		"from":         tr.FromISO,
		"to":           tr.ToISO,
		"kpis":         kpis,
	})
}

// PreConsent returns the legal-data-gap report: how many pageviews
// happened before any consent decision (or with no consent at all).
//
// GET /api/sites/{siteID}/cookies/analytics/pre-consent?from=&to=
func (h *CookieAnalyticsHandler) PreConsent(w http.ResponseWriter, r *http.Request) {
	siteID, ok := h.available(w, r)
	if !ok {
		return
	}
	tr := parseTimeWindow(r)
	share, err := h.analyticsMgr.PreConsentTrafficShare(r.Context(), siteID, tr)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Failed to compute pre-consent share: "+err.Error())
		return
	}
	writeJSON(w, http.StatusOK, share)
}

// TimeToConsent reports P50 / P90 / mean seconds from first pageview to
// first consent decision. Lower = banner is doing its job.
//
// GET /api/sites/{siteID}/cookies/analytics/time-to-consent?from=&to=
func (h *CookieAnalyticsHandler) TimeToConsent(w http.ResponseWriter, r *http.Request) {
	siteID, ok := h.available(w, r)
	if !ok {
		return
	}
	tr := parseTimeWindow(r)
	t, err := h.analyticsMgr.TimeToConsent(r.Context(), siteID, tr)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Failed to compute time-to-consent: "+err.Error())
		return
	}
	writeJSON(w, http.StatusOK, t)
}

// EngagedAcceptRate splits visitors into engaged vs not-engaged cohorts
// and reports accept-all rates.
//
// GET /api/sites/{siteID}/cookies/analytics/engaged-accept-rate?from=&to=
func (h *CookieAnalyticsHandler) EngagedAcceptRate(w http.ResponseWriter, r *http.Request) {
	siteID, ok := h.available(w, r)
	if !ok {
		return
	}
	tr := parseTimeWindow(r)
	er, err := h.analyticsMgr.EngagedAcceptRate(r.Context(), siteID, tr)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Failed to compute engaged accept rate: "+err.Error())
		return
	}
	writeJSON(w, http.StatusOK, er)
}

// TopAcceptingPages returns the entry-page paths with the highest
// accept-all rate (visitors >= 10).
//
// GET /api/sites/{siteID}/cookies/analytics/top-accepting-pages?from=&to=&limit=
func (h *CookieAnalyticsHandler) TopAcceptingPages(w http.ResponseWriter, r *http.Request) {
	siteID, ok := h.available(w, r)
	if !ok {
		return
	}
	tr := parseTimeWindow(r)
	limit := 10
	if v := r.URL.Query().Get("limit"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 && n <= 100 {
			limit = n
		}
	}
	rows, err := h.analyticsMgr.TopAcceptingPages(r.Context(), siteID, tr, limit)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Failed to compute top accepting pages: "+err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"limit": limit,
		"rows":  rows,
	})
}

// DailySplit returns the per-day stacked-bar data for the analytics tab.
//
// GET /api/sites/{siteID}/cookies/analytics/daily?from=&to=
func (h *CookieAnalyticsHandler) DailySplit(w http.ResponseWriter, r *http.Request) {
	siteID, ok := h.available(w, r)
	if !ok {
		return
	}
	tr := parseTimeWindow(r)
	rows, err := h.analyticsMgr.ConsentDailySplit(r.Context(), siteID, tr)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Failed to fetch daily split: "+err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"rows": rows})
}

// VisitorJourney returns a single visitor's stitched timeline. The
// fingerprint comes from the URL path; the time window is the same
// ?from=&to= used elsewhere.
//
// GET /api/sites/{siteID}/cookies/analytics/journey/{fingerprint}?from=&to=
func (h *CookieAnalyticsHandler) VisitorJourney(w http.ResponseWriter, r *http.Request) {
	siteID, ok := h.available(w, r)
	if !ok {
		return
	}
	fp := strings.TrimSpace(urlParam(r, "fingerprint"))
	if fp == "" || len(fp) > 64 {
		writeError(w, http.StatusBadRequest, "Invalid fingerprint")
		return
	}
	tr := parseTimeWindow(r)
	steps, err := h.analyticsMgr.VisitorJourney(r.Context(), siteID, fp, tr)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Failed to fetch visitor journey: "+err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"site_id":     siteID,
		"fingerprint": fp,
		"steps":       steps,
	})
}

File: frontend/src/lib/api/cookieAnalytics.ts
// Frontend API helper for the stitched cookie-analytics surface that the
// Go server exposes via DuckDB ATTACH on the SQLite file. Each function
// matches one /api/sites/{siteID}/cookies/analytics/* endpoint.
//
// All endpoints accept an optional time window via from/to (RFC3339 or
// unix-millis). Defaults to the last 30 days when omitted.

import { apiGet } from './client';

function buildQuery(params: { [key: string]: unknown }): string {
	const usp = new URLSearchParams();
	for (const [k, v] of Object.entries(params)) {
		if (v === undefined || v === null || v === '') continue;
		usp.set(k, String(v));
	}
	const s = usp.toString();
	return s ? `?${s}` : '';
}

export interface FunnelKPIs {
	unique_visitors: number;
	consenting_visitors: number;
	accept_all_visitors: number;
	reject_all_visitors: number;
	custom_visitors: number;
	gpc_visitors: number;
	do_not_sell_visitors: number;
	identified_visitors: number;
	abandoned_banner_visitors: number;
}

export interface FunnelResponse {
	site_id: string;
	from: string;
	to: string;
	kpis: FunnelKPIs;
}

export function funnel(siteID: string, params: { from?: string; to?: string } = {}): Promise<FunnelResponse> {
	return apiGet<FunnelResponse>(`/sites/${siteID}/cookies/analytics/funnel${buildQuery({ ...params })}`);
}

export interface PreConsentShare {
	total_pageviews: number;
	pageviews_before_or_no_consent: number;
	pageviews_after_consent: number;
	pageviews_after_reject: number;
}

export function preConsent(siteID: string, params: { from?: string; to?: string } = {}): Promise<PreConsentShare> {
	return apiGet<PreConsentShare>(`/sites/${siteID}/cookies/analytics/pre-consent${buildQuery({ ...params })}`);
}

export interface TimeToConsent {
	sample_size: number;
	p50_seconds: number;
	p90_seconds: number;
	mean_seconds: number;
}

export function timeToConsent(siteID: string, params: { from?: string; to?: string } = {}): Promise<TimeToConsent> {
	return apiGet<TimeToConsent>(`/sites/${siteID}/cookies/analytics/time-to-consent${buildQuery({ ...params })}`);
}

export interface EngagedAcceptRate {
	engaged_visitors: number;
	engaged_accepted: number;
	engaged_accept_rate_pct: number;
	not_engaged_visitors: number;
	not_engaged_accepted: number;
	not_engaged_accept_rate_pct: number;
}

export function engagedAcceptRate(siteID: string, params: { from?: string; to?: string } = {}): Promise<EngagedAcceptRate> {
	return apiGet<EngagedAcceptRate>(`/sites/${siteID}/cookies/analytics/engaged-accept-rate${buildQuery({ ...params })}`);
}

export interface AcceptingPage {
	path: string;
	visitors: number;
	accepted: number;
	accept_rate_pct: number;
}

export interface TopAcceptingPagesResponse {
	limit: number;
	rows: AcceptingPage[] | null;
}

export function topAcceptingPages(siteID: string, params: { from?: string; to?: string; limit?: number } = {}): Promise<TopAcceptingPagesResponse> {
	return apiGet<TopAcceptingPagesResponse>(`/sites/${siteID}/cookies/analytics/top-accepting-pages${buildQuery({ ...params })}`);
}

export interface ConsentDayRow {
	day: string;
	accept_all: number;
	reject_all: number;
	custom: number;
	gpc: number;
	dns: number;
	do_not_sell: number;
}

export interface DailySplitResponse {
	rows: ConsentDayRow[] | null;
}

export function dailySplit(siteID: string, params: { from?: string; to?: string } = {}): Promise<DailySplitResponse> {
	return apiGet<DailySplitResponse>(`/sites/${siteID}/cookies/analytics/daily${buildQuery({ ...params })}`);
}

export interface VisitorJourneyStep {
	ts_ms: number;
	kind: 'pageview' | 'engagement' | 'consent' | 'identified';
	path?: string;
	method?: string;
	categories?: string;
	gpc_active?: boolean;
	scroll_pct?: number;
	time_on_page_seconds?: number;
}

export interface VisitorJourneyResponse {
	site_id: string;
	fingerprint: string;
	steps: VisitorJourneyStep[];
}

export function visitorJourney(siteID: string, fingerprint: string, params: { from?: string; to?: string } = {}): Promise<VisitorJourneyResponse> {
	return apiGet<VisitorJourneyResponse>(`/sites/${siteID}/cookies/analytics/journey/${encodeURIComponent(fingerprint)}${buildQuery({ ...params })}`);
}
